### ISSUE TITLE
Setup continous integration pipeline

### ISSUE THREAD

**Issue by manoranjith** [direct-state-transfer/dst-go#49]
<!-- Provide a general summary of the task in the title above.

When adding program output or code snippets in the issue, use code blocks with
appropriate syntax highlighting. -->

#### Motivation / Context
<!-- Tell us why this task is important. -->
To setup a continuous integration pipeline.

#### Description                                                                                                                                                  
<!-- Describe the task briefly. -->                                                                                                                               
The CI jobs can be setup on a [circleci](https://circleci.com/) platform. It has a free tier for open source projects. 

- [ ] Configure the pipeline for three jobs:                                                                                             
~1. Checkout.~ Fetching the dependencies directly during lint / test job results in a faster pipeline than using a dedicated container for tests. This is mainly due to the time taken (~15 to 20s) to spin a a dedicated container to fetch dependencies.
2. Run linter.                                                                                                                                                    
3. Run tests.                                                                                                                                                     

- [ ] In the repository, configure the jobs to be  triggered on every push operation on any branch.

- [ ] Display a badge in Readme.md file for CI pipeline status of `master` and `develop` branches.                                                                             
                                                                                                                                                                  
##### Relates to                                                                                                                                                  
<!-- Add links to any related issue or other references here. -->                                                                                                 
Depends on #48.
                                                                                                                                                                  
#### Testing                                                                                                                                                      
<!-- Describe what should be tested and how it can be done. -->                                                                                                   
-NA-
                                                                                                                                                                  
#### Thread Safety                                                                                                                                                
<!-- Should any component of the implementation be thread safe? If so, why? -->                                                                                   
-NA-

**Comment by manoranjith** [comment]
Add a script for checking if all the files contain a valid copyright notice and configure it to run in the lint job.

**Comment by choeppler** [comment]
One thought on that: we could check out the FSFE's [reuse spec](https://reuse.software/) and [tool](https://github.com/fsfe/reuse-tool).

**Comment by manoranjith** [comment]
I will have a look into it.

**Comment by manoranjith** [comment]
As i was not actively looking for tools do this, i added a bash script in the current pull request (#53). The output from it in case of mismatch was only a raw diff output. So it could be useful if this tool could give a better output.

Also it handles other aspects related to compliance looks interesting.

**Comment by choeppler** [comment]
Don't worry in the context of this issue. I was going to look into it anyway and will open a separate issue if necessary...

**Comment by manoranjith** [comment]
The tool was super easy to use. I just had to do `pip3 install --user reuse` and run the below command
```
find . -path "./internal/mocks/*" -prune -o -name "*.go" -exec reuse addheader --license Apache-2.0 {} +
```

Regarding files with build tags, at the first attempt it puts the license header above the build tag, but if we fix it once, it doesn't modify the file in consequent runs. It leaves the build tag above the license header.

**Comment by choeppler** [comment]
Actually, according to the go documentation (cf. https://golang.org/pkg/go/build/), build constraints, a.k.a. build tags, "must appear near the top of the file, preceded only by blank lines and other line comments." So, license and copyright information could precede build tags, couldn't it?

**Comment by manoranjith** [comment]
I looked into the docs again. The build constraint can indeed be preceded by copyright notice. I will update the pull request accordingly.

**Comment by manoranjith** [comment]
I have updated the pull request to look for copyright header at the top of the file. Only generated files, that contain the phrase "DO NOT EDIT" are omitted.
